test: coinselection edge cases

S3RK opened the pull request:
This PR adds tests for some edge cases in coin selection.

There are 4 sections by the type of utxo pool:
1) a few small inputs of equal value
2) one big input and many small inputs
3) just one input
4) one big and one small input

S3RK committed:
test: coinselection edge cases

Xekyo commented on test/functional/wallet_coinselection.py:
```diff
@@ -0,0 +1,261 @@
+        # 1) knapsack latches to the exact match solution and can't find the best solution
+        # SRD can find a solution better than knapsack if draws the big coin in the first 21 tries
+        # chances for SRD to find it are roughly 2*(21/102) = ±40%.
+        # Chances are doubled due to coinselection run twice with and without APS enabled
```
What's "APS"?

jonatack replied:
> What's "APS"?

In more detail, see the configuration flag `-maxapsfee` added by #14582 in 0.21, which sets the max allowed avoid partial spends (APS) fee. It defaults to 0 (fee is the same with and without APS), and setting it to -1 will disable APS, unless `-avoidpartialspends` is set.

Xekyo commented on test/functional/wallet_coinselection.py:
```diff
@@ -0,0 +1,261 @@
+
+        # 1) knapsack latches to the exact match solution and can't find the best solution
+        # SRD can find a solution better than knapsack if draws the big coin in the first 21 tries
+        # chances for SRD to find it are roughly 2*(21/102) = ±40%.
```
I think the chance is (1-(101/102)×(100/101)×…×(81/82))=(1-81/102)… oh, I see, that's (21/102). So, to get a better SRD solution that would need to occur at least once, which is 1-(81/102)² = 36.93%… okay, I gotcha.

S3RK replied:
Yes. that is correct. Not sure if I can improve the comment here somehow

Xekyo commented on test/functional/wallet_coinselection.py:
```diff
@@ -0,0 +1,261 @@
+        assert_approx(max_fee, best_fee)
+        assert_approx(fee_total, best_fee * 100)
+
+        # 1) knapsack latches to the exact match solution and can't find the best solution
```
We've previously used "exact match" for any changeless solution, so it wasn't immediately obvious to me that it was used in the literal sense here. Perhaps you should clarify that you mean that it's literally the same to the sat.

S3RK replied:
Will update the comment

S3RK replied:
Updated, but I think we should stop using "exact match" in place of changeless solution. Changeless solution is a solution in the range around exact match constrained by the cost of change/discard fee rate.

Xekyo commented on test/functional/wallet_coinselection.py:
```diff
@@ -0,0 +1,261 @@
+        assert_greater_than(knapsack_fee * 100, fee_total)
+        assert_greater_than(fee_total, best_fee * 100)
+
+        # 2) no exact match, effective value of the big coin is smaller than target + MIN_CHANGE
```
I realized above use of "exact match" here. :)

S3RK replied:
Thanks. Will retouch the comment

Xekyo commented on test/functional/wallet_coinselection.py:
```diff
@@ -0,0 +1,261 @@
+        assert_greater_than(fee_total, best_fee * 100)
+
+        # 2) no exact match, effective value of the big coin is smaller than target + MIN_CHANGE
+        # BnB can find the same solution from previous test but dropping the change output to fees
```
```suggestion
        # BnB will find the same solution from previous test but dropping the change output to fees
```
BnB is deterministic, so if it "can" I would expect that it does? Does "change output" mean "remainder" here?—Is this the BnB without an upper bound or the variant we currently use?

S3RK replied:
Any BnB, since we're within the cost of change bound. I'll retouch the comment

Xekyo commented on test/functional/wallet_coinselection.py:
```diff
@@ -0,0 +1,261 @@
+
+        # 4) knapsack latches to exact match + MIN_CHANGE
+        # SRD still occasionally find better solution
+        fee_total, max_fee = select_coins(big_coin=1.00051, target=0.99, iterations=100)
```
It's a bit hard for me to follow how the exact values were picked, since both the `big_coin` and `target` are changing. I haven't tried to step by step reproduce each calculation in this first review, and I do see that many of your test cases do explain some of the calculation, but I just noticed that this one for example does not.

Since it's not trivial for me to follow, I kinda expect that other reviewers might also need a bit help. :grin:

S3RK replied:
This is basically the same as test case "1)" in this section, but all the values decreased by MIN_CHANGE. The idea is to  demonstrate that Knapsack latches not only to down-to-satoshi matches. I'll improve the comment

Xekyo commented on test/functional/wallet_coinselection.py:
```diff
@@ -0,0 +1,261 @@
+        assert_equal(len(tx['vin']), 1)
+        assert_equal(len(tx['vout']), 2)
+
+        # if small coins > MIN_CHANGE than we just add one small coin to the big one and create change
```
This comment confuses me. After paying the fees, the big coin doesn't have enough funds to pay fees and create the output here, so we would always need both coins.

S3RK replied:
Oops. The idea was that you have more than 1 small coin. If small coins > MIN_CHANGE that the tests in `test_one_big_and_many_small_coins` doesn't work. Any idea how to improve that? For now, added more than one small coin in this test case

murchandamus reviewed (commented):
Cool! Concept ACK. Did a light review, haven't reproduced all the numbers.

S3RK commented on test/functional/wallet_coinselection.py:
```diff
@@ -0,0 +1,261 @@
+        # We will target a solution with 20 coins that will add up exactly to 1btc.
+        # Add a coin that will cover all the fees and exactly match the target.
+        # TODO: fix that knapsack targets change even if there is no change needed
+        # (148*20+42+31)*350+5000000 = 6061550
```
42 is header (11) + p2pwkh (31). I'll updated the comment

S3RK commented on test/functional/wallet_coinselection.py:
```diff
@@ -0,0 +1,261 @@
+        # assert_equal(len(tx['vout']), 1)
+        # assert(tx['vin'][0]['txid'] in [tx1, tx2])
+
+        # fees for 1 legacy input tx = (148+42)*20 = 3800sat
```
Yes, 42 is output + header. I'll update the comment

S3RK commented on test/functional/wallet_coinselection.py:
```diff
@@ -0,0 +1,261 @@
+            result = w1.fundrawtransaction(w1.createrawtransaction([], {dummyaddress: target}), {"fee_rate": fee_rate})
+            return w1.decoderawtransaction(result['hex'])
+
+        # 1-input-1-output tx is 110vbytes which results in 2200sat fee at 20sat/vbyte
```
I'll update the comment

S3RK commented on test/functional/wallet_coinselection.py:
```diff
@@ -0,0 +1,261 @@
+        assert_equal(tx['vin'][0]['txid'], tx3)
+
+        # a solution with two bech32 inputs and a change is the most efficient
+        # but it could only be found by SRD
```
Exactly. Knapsack prefers a solution with legacy+bech32 as it's closer to the target + MIN_CHANGE due to higher fees.
I'll update the comment

S3RK commented on test/functional/wallet_coinselection.py:
```diff
@@ -0,0 +1,261 @@
+            return fee_total, max_fee
+
+        # in all the scenarios below the best solution would be to take one big coin and create a change output
+        best_fee = (42 + 68 + 31) * fee_rate / COIN
```
yes, it's. I'll split off the calculation

S3RK commented on test/functional/wallet_coinselection.py:
```diff
@@ -0,0 +1,261 @@
+
+        # We will target a solution with 20 coins that will add up exactly to 1btc.
+        # Add a coin that will cover all the fees and exactly match the target.
+        # TODO: fix that knapsack targets change even if there is no change needed
```
I'm not really arguing for improving it. I guess maybe one way that fix that would be to just remove Knapsack that targets no change and leave only knapsack that targets a MIN_CHANGE.

P.S. BnB won't produce same solutions even in the situations when a changeless solution is possible, because BnB optimizes for waste. But I think this is okay.

S3RK commented on test/functional/wallet_coinselection.py:
```diff
@@ -0,0 +1,261 @@
+        # 1) 2x bech32 solution (SRD only)
+        # 2) legacy + bech32 solution
+        tx = select_coins_for(0.99994899)
+        # legacy + dropping more for fees is still better in terms of waste than legacy + bech32
```
To keep the solution with best waste out of a 1000 trials? Hm.. that would fix this test case, but it'd make Knapsack behave exactly like running SRD in a loop. I guess we can then just drop knapsack altogether.

S3RK commented on test/functional/wallet_coinselection.py:
```diff
@@ -0,0 +1,261 @@
+        w1 = self.nodes[0].get_wallet_rpc("test_one_big_and_many_small_coins")
+
+        # Create a bunch of small coins
+        for i in range(100):
```
Actually I was trying to keep the number of small coins as a low as possible while still demonstrating some inefficiencies of coin selection.
1) The more coins we have -> the more time the test takes.
2) The more coins we have -> the harder it's to debug and to reason about.

What is the test scenario that would be possible with 1000 coins, but not possible with 100 coins?

P.S. I think we can definitely reproduce this scenario, I see no reasons why not? If utxo pool is equivalent to this test cases then even with SRD it will find better solution in ~856/1000 = 85%, which leaves 15% to a really bad knapsack alternative.

Xekyo replied:
Yeah, I think it would be easy to produce the scenario: 

• 100 small coins, 1 big coin
• no BnB solution
• a large set of the small coins produces a smaller change output than the 1 big coin, so knapsack prefers the former
• SRD only finds the 1 big coin by itself as a solution in ~1/101 cases

⇒ usually a worse waste than best solution, but better than knapsack

S3RK replied:
Isn't this exactly #3 in `test_one_big_and_many_small_coins`?

S3RK commented on test/functional/wallet_coinselection.py:
```diff
@@ -0,0 +1,261 @@
+        tx3 = w0.sendtoaddress(w1.getnewaddress("", "legacy"), 1)
+        self.generate(self.nodes[0], 1)
+
+        # TODO: set long_term_fee_rate explicitly
```
fixed that

S3RK commented:
Retouched comments to address Xekyo's feedback.

Xekyo commented:
Looks good. What is the next step you plan for this PR? How can one help?

S3RK commented:
I thought this PR can be merged roughly as it is (addressing all the concerns of course), as I believe the test is valuable in itself.

As a follow up I'd like to fix some of the inefficiencies of coin selection by increasing BnB upper limit. But for that we need to improve how we drop change for fees.  I tried one approach in #23367 but it doesn't work that well. So I'm going splitting #23367 into parts that could be merged independently (the first one -> #24649).

Let's talk on the next wallet meeting.

S3RK commented:
There is probably a conflict with #24494.

achow101 commented:
While I appreciate more tests, these tests seem to be extremely specific to the way that coin selection currently works and look like they would start failing as soon as we change something, such as removing knapsack. So I'm not sure if we actually want to have these tests if they are going to be removed and/or require a ton of recalculation for future coin selection changes.

If we do want these tests, then I don't think it should use legacy addresses so much as hopefully we won't make those by default soon.

S3RK commented:
> While I appreciate more tests, these tests seem to be extremely specific to the way that coin selection currently works and look like they would start failing as soon as we change something, such as removing knapsack. So I'm not sure if we actually want to have these tests if they are going to be removed and/or require a ton of recalculation for future coin selection changes. 

Agree, these are not your normal tests, primarily because it's hard to define the "right" or "expected" outcome. I created these tests to expose some shortcomings of current coin selection and justify increasing BnB limit (#24752). It'd be nice to have them around in some form to avoid regressions, maybe we can incorporate them somehow into simulations. wdyt?

> If we do want these tests, then I don't think it should use legacy addresses so much as hopefully we won't make those by default soon.

instagibbs commented:
To throw my relatively unproductive opinion into the mix, this is a staggeringly large number of magic numbers. Ideally all numbers in tests are "obvious" or programmatic. It's a nightmare to maintain otherwise.

Didn't read or think enough about overall thrust of the tests.

S3RK commented:
Thanks for the feedback. Maybe a better way to verify coin selection would to integrate some simulation/benchmark into this repo.
